libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp

```diff
@@ -461,35 +471,49 @@ bool NavEKF3_core::readyToUseBodyOdm(void) const
 // return true if the filter to be ready to use gps
 bool NavEKF3_core::readyToUseGPS(void) const
 {
-    return validOrigin && tiltAlignComplete && yawAlignComplete && (delAngBiasLearned || assume_zero_sideslip()) && gpsGoodToAlign && (frontend->_fusionModeGPS != 3) && gpsDataToFuse && !gpsInhibit;
+    if (frontend->_sources.getPosXYSource() != AP_NavEKF_Source::SourceXY::GPS) {
+        return false;
+    }
+
+    return validOrigin && tiltAlignComplete && yawAlignComplete && (delAngBiasLearned || assume_zero_sideslip()) && gpsGoodToAlign && gpsDataToFuse && !gpsInhibit;
 }
 
 // return true if the filter to be ready to use the beacon range measurements
 bool NavEKF3_core::readyToUseRangeBeacon(void) const
 {
+    if (frontend->_sources.getPosXYSource() != AP_NavEKF_Source::SourceXY::BEACON) {
+        return false;
+    }
+
     return tiltAlignComplete && yawAlignComplete && delAngBiasLearned && rngBcnAlignmentCompleted && rngBcnDataToFuse;
 }
 
 // return true if the filter is ready to use external nav data
 bool NavEKF3_core::readyToUseExtNav(void) const
 {
+    if (frontend->_sources.getPosXYSource() != AP_NavEKF_Source::SourceXY::EXTNAV) {
+        return false;
+    }
+
     return tiltAlignComplete && extNavDataToFuse;
 }
 
 // return true if we should use the compass
 bool NavEKF3_core::use_compass(void) const
 {
     const auto *compass  = dal.get_compass();
-    return effectiveMagCal != MagCal::EXTERNAL_YAW &&
-           compass &&
-           compass->use_for_yaw(magSelectIndex) &&
-           !allMagSensorsFailed;
+    const AP_NavEKF_Source::SourceYaw yaw_source = frontend->_sources.getYawSource();
+    return (yaw_source == AP_NavEKF_Source::SourceYaw::COMPASS || yaw_source == AP_NavEKF_Source::SourceYaw::EXTERNAL_COMPASS_FALLBACK) &&
+            compass &&
+            compass->use_for_yaw(magSelectIndex) &&
+            !allMagSensorsFailed;
```

Comment:
I was trying to keep the substitution as close as possible to one-to-one so even where I saw improvements i resisted making them because all it would do is slow the review.

Still, because it's your suggestion I've changed it, thanks!